### DISCUSSION TITLE
Declaring empty external functions

### DISCUSSION THREAD

**Discussion by jmintb** [mlir-rs/melior#291] [Q&A] [answered]
Hi 

I am attempting to call printf and I believe that requires me to declare it, so something like this:

```mlir
func @printf(i8)
```
At least that is my interpretation from https://mlir.llvm.org/docs/Dialects/Func/#funcfunc-funcfuncop.

I am a bit stuck on how to do that with melior. I have tried a few things with `melior::dialect::func::func` with no success.

Thanks for the awesome crate!

TLDR: How can I declare external functions?

**Comment by raviqqe** [comment, marked as answer]
You can pass an empty region as a body of the `func.func` operation (i.e. `Region::new()`.) BTW, the example in the documentation seems to be wrong as FAQ and symbol visibility documentation actually says external function definitions must have `private` visibility. https://mlir.llvm.org/getting_started/Faq/#whats-with-op-symbol-declaration-cannot-have-public-visibility

See also #292.

**Reply by jmintb** [reply]
Awesome thanks!
